feat: initial info command for tools

maxandersen opened the pull request:
initial version of `jbang info <script>`.

currently dumps a json doc like:

```
 jbang info examples/smee.java
{
  "originalResource": "examples/smee.java",
  "backingResource": "examples/smee.java",
  "resolvedClasspath": [
    "/Users/max/.m2/repository/info/picocli/picocli/4.5.0/picocli-4.5.0.jar",
    "/Users/max/.m2/repository/org/jboss/resteasy/resteasy-client/4.4.1.Final/resteasy-client-4.4.1.Final.jar",
    "/Users/max/.m2/repository/org/jboss/resteasy/resteasy-client-api/4.4.1.Final/resteasy-client-api-4.4.1.Final.jar",
    "/Users/max/.m2/repository/org/jboss/resteasy/resteasy-core-spi/4.4.1.Final/resteasy-core-spi-4.4.1.Final.jar",
    "/Users/max/.m2/repository/org/jboss/spec/javax/xml/bind/jboss-jaxb-api_2.3_spec/1.0.1.Final/jboss-jaxb-api_2.3_spec-1.0.1.Final.jar",
    "/Users/max/.m2/repository/org/reactivestreams/reactive-streams/1.0.3/reactive-streams-1.0.3.jar",
    "/Users/max/.m2/repository/jakarta/validation/jakarta.validation-api/2.0.2/jakarta.validation-api-2.0.2.jar",
    "/Users/max/.m2/repository/org/jboss/spec/javax/annotation/jboss-annotations-api_1.3_spec/2.0.1.Final/jboss-annotations-api_1.3_spec-2.0.1.Final.jar",
    "/Users/max/.m2/repository/com/sun/activation/jakarta.activation/1.2.1/jakarta.activation-1.2.1.jar",
    "/Users/max/.m2/repository/org/jboss/resteasy/resteasy-core/4.4.1.Final/resteasy-core-4.4.1.Final.jar",
    "/Users/max/.m2/repository/org/eclipse/microprofile/config/microprofile-config-api/1.3/microprofile-config-api-1.3.jar",
    "/Users/max/.m2/repository/org/osgi/org.osgi.annotation.versioning/1.0.0/org.osgi.annotation.versioning-1.0.0.jar",
    "/Users/max/.m2/repository/io/smallrye/smallrye-config/1.3.9/smallrye-config-1.3.9.jar",
    "/Users/max/.m2/repository/javax/enterprise/cdi-api/2.0.SP1/cdi-api-2.0.SP1.jar",
    "/Users/max/.m2/repository/javax/el/javax.el-api/3.0.0/javax.el-api-3.0.0.jar",
    "/Users/max/.m2/repository/javax/interceptor/javax.interceptor-api/1.2/javax.interceptor-api-1.2.jar",
    "/Users/max/.m2/repository/javax/inject/javax.inject/1/javax.inject-1.jar",
    "/Users/max/.m2/repository/org/jboss/logging/jboss-logging/3.3.2.Final/jboss-logging-3.3.2.Final.jar",
    "/Users/max/.m2/repository/org/apache/httpcomponents/httpclient/4.5.10/httpclient-4.5.10.jar",
    "/Users/max/.m2/repository/org/apache/httpcomponents/httpcore/4.4.12/httpcore-4.4.12.jar",
    "/Users/max/.m2/repository/commons-logging/commons-logging/1.2/commons-logging-1.2.jar",
    "/Users/max/.m2/repository/commons-codec/commons-codec/1.13/commons-codec-1.13.jar",
    "/Users/max/.m2/repository/commons-io/commons-io/2.5/commons-io-2.5.jar",
    "/Users/max/.m2/repository/org/jboss/spec/javax/ws/rs/jboss-jaxrs-api_2.1_spec/2.0.1.Final/jboss-jaxrs-api_2.1_spec-2.0.1.Final.jar",
    "/Users/max/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.2.3/jackson-databind-2.2.3.jar",
    "/Users/max/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.2.3/jackson-annotations-2.2.3.jar",
    "/Users/max/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.2.3/jackson-core-2.2.3.jar"
  ]
}

```


<!--
Thanks for submitting your Pull Request!

Please delete this text, and add description of the topic solved by this PR.

To make releases as automated as possible we use conventional commits formats.
Thus commits and pull-requests titles should before merge follow the Conventional Commits spec.

Details in https://github.com/zeke/semantic-pull-requests

If in doubt then open the pull-request and we'll help you - Thank you!
-->

maxandersen commented:
/cc @fbricon

fbricon commented:
Can you add the Java config as well?

fbricon commented:
Will also prolly need to report resolution errors, eg //DEP foo:bar:1.0, if can't be resolved, then we should see the error message bound to the location of the DEP declaration

maxandersen commented:
Java config As in Java version or the exact jdk location?

Resolution issues is doable - just wondering how to represent it.

fbricon commented:
Will need java version (eg 15) + options (-parameters --enable-preview)

As for the resolution issues, maybe something like:
```json
{
  "errors": [
      {
        "originalResource":"file.java",   // or use backingResource, what's the difference between the 2?
        "message":"foo:bar:1.0 can not be resolved"
        "stacktrace": "...stacktrace..."
        "location": {
             line: 3, // that's assuming errors only cover 1 line
             start: 4,  // start column
             end: 25. // end column
         }
     }
  ]
 }
```

maxandersen commented:
> ```json
>         "originalResource":"file.java",   // or use backingResource, what's the difference between the 2?
> ```

there are a few cases where they will differ:

1) jbang https://some/path/to/max.java then backingResource will be where that url is downloaded to
2) jbang my-plugin.sh then backingresource will be a symbolic linked or even copy named `MyPlugin.java`

...but except for #2 not sure tools will ever see it as you don't (yet) have a way to "browse" a url ;)

fbricon commented:
would be nice if there was a flag to also resolve sources in the same request

maxandersen commented:
here is an update that gives a better command structure for info command.

@fbricon for you it means you wanna call `jbang info tools <script>` now.

I've added a `jbang info classpath` command too which prints out the resolved classpath to use for runs. its still missing the actually generated jar...but that will come.

fbricon commented:
> Task :spotlessJavaCheck FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':spotlessJavaCheck'.
> The following files had format violations:
      src/main/java/dev/jbang/cli/Info.java
          @@ -1,15 +1,17 @@
           package·dev.jbang.cli;

          -import·com.google.gson.Gson;
          -import·com.google.gson.GsonBuilder;
          -import·dev.jbang.Script;
          -import·picocli.CommandLine;
          +import·static·dev.jbang.Settings.CP_SEPARATOR;

           import·java.io.IOException;
           import·java.util.Arrays;
           import·java.util.List;

          +import·com.google.gson.Gson;
          +import·com.google.gson.GsonBuilder;
          +
          +import·dev.jbang.Script;
          +
          -import·static·dev.jbang.Settings.CP_SEPARATOR;
          +import·picocli.CommandLine;

           @CommandLine.Command(name·=·"info",·description·=·"Provides·info·about·the·script·for·tools·(and·humans·who·are·tools).",·subcommands·=·{
           \t\tTools.class,·ClassPath.class·})
  Run './gradlew :spotlessApply' to fix these violations.

maxandersen commented:
refresh.

maxandersen commented:
@fbricon merged in current work so next release will have this but i don't consider the "format" stable yet.

I changed "resolvedClasspath" to "resolvedDependencies" as it is just the dependencies - not containing the classpath for the built app.

i also added javaVersion which will for now be in the format of what //JAVA_VERSION  honors. so 14 or 14+ for example. we should probably "resolve" this and also return resolved jdk path.

your other requests I'll take on as separate issues as i can see will need to redo some of the command line generation we do at the moment to return the info tools will work best with. ie.. I can't currently easily distingish which part of the command line is "additional args" and which are just the ones neede to run.

quintesse commented:
> we should probably "resolve" this and also return resolved jdk path.

You can use `JdkManager.getCurrentJdk(version)` for that. It'll give you the Path to the JDK that matches the version string. (Btw, this _will_ download and install a JDK if necessary).
